dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -177,31 +179,78 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
 
     winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveByIdAsync(uint32_t notificationId)
     {
-        throw hresult_not_implemented();
+        THROW_HR_IF(E_INVALIDARG, notificationId == 0);
+
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        ToastNotifications_RemoveToast(appId.c_str(), notificationId);
     }
 
     winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveWithTagAsync(hstring tag)
     {
-        throw hresult_not_implemented();
+        std::wstring _tag{ tag.c_str() };
+
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        ToastNotifications_RemoveToastsWithTagAndGroup(appId.c_str(), _tag.c_str(), nullptr);
     }
 
     winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveWithTagGroupAsync(hstring tag, hstring group)
     {
-        throw hresult_not_implemented();
+        std::wstring _tag{ tag.c_str() };
+        std::wstring _group{ group.c_str() };
+
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        ToastNotifications_RemoveToastsWithTagAndGroup(appId.c_str(), _tag.c_str(), _group.c_str());
     }
 
     winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveGroupAsync(hstring group)
     {
-        throw hresult_not_implemented();
+        std::wstring _group{ group.c_str() };
+
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        ToastNotifications_RemoveToastsWithTagAndGroup(appId.c_str(), nullptr, _group.c_str());
     }
 
     winrt::Windows::Foundation::IAsyncAction AppNotificationManager::RemoveAllAsync()
     {
-        throw hresult_not_implemented();
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        THROW_IF_FAILED(ToastNotifications_RemoveAllToastsForApp(appId.c_str()));
     }
 
     winrt::Windows::Foundation::IAsyncOperation<winrt::Windows::Foundation::Collections::IVector<winrt::Microsoft::Windows::AppNotifications::AppNotification>> AppNotificationManager::GetAllAsync()
     {
-        throw hresult_not_implemented();
+        co_await winrt::resume_background();
+
+        std::wstring appId{ RetrieveNotificationAppId() };
+        winrt::com_ptr<ToastABI::IVector<ToastABI::INotificationProperties*>> toastPropertiesCollection;
+        THROW_IF_FAILED(ToastNotifications_GetHistory(appId.c_str(), toastPropertiesCollection.put()));
```

Comment:
Make sure that this doesn't throw if there are no toasts. Otherwise we have a problem. 

Recommend: LOG_IF_FAILED instead.

Reply:
It won't throw if there are no active toasts but it will throw if an internal failure happens. Are you suggesting we should swallow all form of errors?